bigquery/storage/managedwriter/adapt/protoconversion.go

```diff
@@ -0,0 +1,259 @@
+				if err != nil {
+					return nil, fmt.Errorf("couldn't compile (%s): %v", currentScope, err)
+				}
+				// Now that we have the submessage definition, we append it both to the local dependencies, as well
+				// as inserting it into the cache for possible reuse elsewhere.
+				deps = append(deps, desc.ParentFile())
+				err = cache.add(ts, desc)
+				if err != nil {
+					return nil, fmt.Errorf("failed to add descriptor to cache: %v", err)
+				}
+				fdp, err := tableFieldSchemaToFieldDescriptorProto(f, fNumber, currentScope)
+				if err != nil {
+					return nil, fmt.Errorf("couldn't compute field schema (%s): %v", currentScope, err)
+				}
+				fields = append(fields, fdp)
+			}
+		} else {
+			fd, err := tableFieldSchemaToFieldDescriptorProto(f, fNumber, currentScope)
+			if err != nil {
+				return nil, err
+			}
+			fields = append(fields, fd)
+		}
+	}
+	// Start constructing a DescriptorProto.
+	dp := &descriptorpb.DescriptorProto{
+		Name:  proto.String(scope),
+		Field: fields,
+	}
+
+	// Use the local dependencies to generate a list of filenames.
+	depNames := []string{
+		wellKnownTypesWrapperName,
+	}
+	for _, d := range deps {
+		depNames = append(depNames, d.ParentFile().Path())
+	}
+
+	// Now, construct a FileDescriptorProto.
+	fdp := &descriptorpb.FileDescriptorProto{
+		MessageType: []*descriptorpb.DescriptorProto{dp},
+		Name:        proto.String(fmt.Sprintf("%s.proto", scope)),
+		Syntax:      proto.String("proto3"),
+		Dependency:  depNames,
+	}
+
+	// We'll need a FileDescriptorSet as we have a FileDescriptorProto for the current
+	// descriptor we're building, but we need to include all the referenced dependencies.
+	fds := &descriptorpb.FileDescriptorSet{
+		File: []*descriptorpb.FileDescriptorProto{
+			fdp,
+			protodesc.ToFileDescriptorProto(wrapperspb.File_google_protobuf_wrappers_proto),
+		},
+	}
+	for _, d := range deps {
+		fds.File = append(fds.File, protodesc.ToFileDescriptorProto(d))
+	}
+
+	// Load the set into a registry, then interrogate it for the descriptor corresponding to the top level message.
+	files, err := protodesc.NewFiles(fds)
+	if err != nil {
+		return nil, fmt.Errorf("protodesc.NewFiles: %v", err)
+	}
+	return files.FindDescriptorByName(protoreflect.FullName(scope))
+}
+
+// tableFieldSchemaToFieldDescriptorProto builds individual field descriptors for a proto message.
+// We're using proto3 syntax, but BigQuery supports the notion of NULLs which conflicts with proto3 default value
+// behavior.  To enable it, we look for nullable fields in the schema that should be scalars, and use the
+// well-known wrapper types.
+//
+// Messages are always nullable, and repeated fields are as well.
+func tableFieldSchemaToFieldDescriptorProto(field *storagepb.TableFieldSchema, idx int32, scope string) (*descriptorpb.FieldDescriptorProto, error) {
+	name := strings.ToLower(field.GetName())
+	if field.GetType() == storagepb.TableFieldSchema_STRUCT {
```

Comment:
Maybe use a switch here?

Reply:
Perhaps if the logic expands.  Right now the struct case is the significant deviation from the others, which all differ slightly in specific enum values.